src/PAModel/Serializers/SourceSerializer.cs

```diff
@@ -795,5 +797,78 @@ private static BuildVerJson GetBuildDetails()
             }
         }
 
+        private static void ValidateIfTemplateExists(CanvasDocument app, BlockNode node, ErrorContainer errors)
+        {
+            // parse all children of the root nodes to see if al of them have the temeplate they require.
+            foreach(var child in node.Children)
+            {
+                if (string.IsNullOrEmpty(child?.Name?.Kind?.TypeName))
+                {
+                    // could not perform validation, hence continue.
+                    continue;
+                }
+
+                // check if it is a group control. group control isn't a real control and don't have templates, but have child control which have templates.
+                if (child.Name.Kind.TypeName == "group" || child.Name.Kind.TypeName == "groupContainer")
+                {
+                    foreach (var child1 in child.Children)
+                    {
+                        ValidateIfTemplateExistsInternal(app, node, child1, errors);
+                    }
+                }
+                // validate if template exists.
+                else
+                {
+                    ValidateIfTemplateExistsInternal(app, node, child, errors);
+                }
+            }
+        }
+
+        private static void ValidateIfTemplateExistsInternal(CanvasDocument app, BlockNode parentNode, BlockNode childNode, ErrorContainer errors)
+        {
+            if (string.IsNullOrEmpty(childNode?.Name?.Kind?.TypeName))
+            {
+                // could not perform validation.
+                return;
+            }
+            // group and grouContainer controls aren't real controls they contain other child controls.
+            if (childNode.Name.Kind.TypeName == "group" || childNode.Name.Kind.TypeName == "groupContainer")
+            {
+                foreach (var child in childNode.Children)
+                {
+                    ValidateIfTemplateExistsInternal(app, childNode, child, errors);
+                }
+                return;
+            }
+            
+            CombinedTemplateState templateState;
+            app._templateStore.TryGetTemplate(childNode.Name.Kind.TypeName, out templateState);
+            // For each of the template there mus be an entry in the _templateStore.
+            if(templateState == null)
+            {
+                errors.ValidationError($"Could not find the template with name: {childNode.Name.Kind.TypeName} which is referred in {parentNode.Name.Identifier}. " +
+                    $"If the template was deleted intentionally please make sure to update the source files to remove the references to this template.");
+            }
+            // If its a widget template then there must be a xml file in the pkgs directory.
+            if(templateState.IsWidgetTemplate && !app._templates.UsedTemplates.Any(x => x.Name.Equals(childNode.Name.Kind.TypeName, StringComparison.InvariantCultureIgnoreCase)))
```

Comment:
>InvariantCultureIgnoreCase [](http://example.com/codeflow?start=149&length=26)

double checking string compares here -  why is this one IgnoreCase, but the one on line 859 isn't?

Reply:
Removed these ignore case checks, doesn't seem like these are needed.